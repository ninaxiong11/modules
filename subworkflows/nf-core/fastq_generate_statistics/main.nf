//
// Short read sequencing data QC using different tools
//
include { FASTQC       } from '../../../modules/nf-core/fastqc/main'
include { SEQFU_CHECK  } from '../../../modules/nf-core/seqfu/check/main'
include { SEQFU_STATS  } from '../../../modules/nf-core/seqfu/stats/main'
include { SEQKIT_STATS } from '../../../modules/nf-core/seqkit/stats/main'
include { SEQTK_COMP   } from '../../../modules/nf-core/seqtk/comp/main'

workflow FASTQ_GENERATE_STATISTICS {

    take:
    ch_reads          // channel: [ val(meta), [ fastq ] ]
    skip_fastqc       // boolean
    skip_seqfu_check  // boolean
    skip_seqfu_stats  // boolean
    skip_seqkit_stats // boolean
    skip_seqtk_comp   // boolean

    main:

    ch_versions = Channel.empty()

    if (!skip_fastqc) {
        FASTQC( ch_reads )
        ch_versions = ch_versions.mix(FASTQC.out.versions.first())
    }

    if (!skip_seqfu_check){
        SEQFU_CHECK( ch_reads )
        ch_versions = ch_versions.mix(SEQFU_CHECK.out.versions.first())
    }

    if (!skip_seqfu_stats) {
        SEQFU_STATS ( ch_reads )
        ch_versions = ch_versions.mix(SEQFU_STATS.out.versions.first())
    }

    if (!skip_seqkit_stats) {
        SEQKIT_STATS ( ch_reads )
        ch_versions = ch_versions.mix(SEQKIT_STATS.out.versions.first())
    }

    if (!skip_seqtk_comp) {
        SEQTK_COMP ( ch_reads )
        ch_versions = ch_versions.mix(SEQTK_COMP.out.versions.first())
    }

    emit:
    fastqc_html   = FASTQC.out.html
    fastqc_zip    = FASTQC.out.zip
    seqfu_check   = SEQFU_CHECK.out.check
    seqfu_stats   = SEQFU_STATS.out.stats
    seqfu_multiqc = SEQFU_STATS.out.multiqc
    seqkit_stats  = SEQKIT_STATS.out.stats
    seqtk_stats   = SEQTK_COMP.out.seqtk_stats
    versions      = ch_versions
}
