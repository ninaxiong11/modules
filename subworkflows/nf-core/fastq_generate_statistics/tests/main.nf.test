nextflow_workflow {
    name "Test Subworkflow FASTQ_GENERATE_STATISTICS"
    script "../main.nf"
    workflow "FASTQ_GENERATE_STATISTICS"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/fastq_generate_statistics"
    tag "fastqc"
    tag "seqfu"
    tag "seqfu/check"
    tag "seqfu/stats"
    tag "seqkit"
    tag "seqkit/stats"
    tag "seqtk"
    tag "seqtk/comp"

    test("sarscov2 - fastq - single_end") {
        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test_single', single_end:true ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true)
                ])
                input[1] = false
                input[2] = false
                input[3] = false
                input[4] = false
                input[5] = false
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    file(workflow.out.fastqc_html[0][1]).name,
                    file(workflow.out.fastqc_zip[0][1]).name,
                    workflow.out.seqfu_check,
                    workflow.out.seqfu_stats,
                    workflow.out.seqfu_multiqc,
                    workflow.out.seqkit_stats,
                    workflow.out.seqtk_stats,
                    workflow.out.versions
                ).match() }
            )
        }
    }
    test("sarscov2 - fastq - paired_end") {
        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test_paired', single_end:false ], // meta map
                    [
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkIfExists: true)
                    ]
                ])
                input[1] = false
                input[2] = false
                input[3] = false
                input[4] = false
                input[5] = false
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    file(workflow.out.fastqc_html[0][1][0]).name,
                    file(workflow.out.fastqc_zip[0][1][0]).name,
                    file(workflow.out.fastqc_html[0][1][1]).name,
                    file(workflow.out.fastqc_zip[0][1][1]).name,
                    workflow.out.seqfu_check,
                    workflow.out.seqfu_stats,
                    workflow.out.seqfu_multiqc,
                    workflow.out.seqkit_stats,
                    workflow.out.seqtk_stats,
                    workflow.out.versions
                ).match() }
            )
        }
    }
    test("sarscov2 - fastq - single_end - stub") {

        options "-stub"

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test_single', single_end:true ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true)
                ])
                input[1] = false
                input[2] = false
                input[3] = false
                input[4] = false
                input[5] = false
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out,
                    workflow.out.versions.collect{ path(it).yaml }.unique()
                ).match() }
            )
        }
    }
    test("sarscov2 - fastq - paired_end - stub") {

        options "-stub"

        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test_paired', single_end:false ], // meta map
                    [
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true),
                        file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkIfExists: true)
                    ]
                ])
                input[1] = false
                input[2] = false
                input[3] = false
                input[4] = false
                input[5] = false
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out,
                    workflow.out.versions.collect{ path(it).yaml }.unique()
                ).match() }
            )
        }
    }
}
