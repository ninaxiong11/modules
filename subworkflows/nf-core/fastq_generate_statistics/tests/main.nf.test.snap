{
    "sarscov2 - fastq - single_end - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_single",
                            "single_end": true
                        },
                        "test_single.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_single",
                            "single_end": true
                        },
                        "test_single.zip:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test_single",
                            "single_end": true
                        },
                        "test_single.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test_single",
                            "single_end": true
                        },
                        "test_single.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test_single",
                            "single_end": true
                        },
                        "test_single_mqc.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test_single",
                            "single_end": true
                        },
                        "test_single.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test_single",
                            "single_end": true
                        },
                        "test_single.seqtk_stats.tsv:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "7": [
                    "versions.yml:md5,5cd377fecf8c1a6bb9aa657313e882d4",
                    "versions.yml:md5,7518d6130f9f32acd21e5ebaaeb02c82",
                    "versions.yml:md5,b3a5ec7e12cfad248f89471add43438f",
                    "versions.yml:md5,bd055fdc48b00af8df08c1f30575bfd6",
                    "versions.yml:md5,d1c67dc2d4ff10a37e0f6e53e243548e"
                ],
                "fastqc_html": [
                    [
                        {
                            "id": "test_single",
                            "single_end": true
                        },
                        "test_single.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "fastqc_zip": [
                    [
                        {
                            "id": "test_single",
                            "single_end": true
                        },
                        "test_single.zip:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "seqfu_check": [
                    [
                        {
                            "id": "test_single",
                            "single_end": true
                        },
                        "test_single.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "seqfu_multiqc": [
                    [
                        {
                            "id": "test_single",
                            "single_end": true
                        },
                        "test_single_mqc.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "seqfu_stats": [
                    [
                        {
                            "id": "test_single",
                            "single_end": true
                        },
                        "test_single.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "seqkit_stats": [
                    [
                        {
                            "id": "test_single",
                            "single_end": true
                        },
                        "test_single.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "seqtk_stats": [
                    [
                        {
                            "id": "test_single",
                            "single_end": true
                        },
                        "test_single.seqtk_stats.tsv:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,5cd377fecf8c1a6bb9aa657313e882d4",
                    "versions.yml:md5,7518d6130f9f32acd21e5ebaaeb02c82",
                    "versions.yml:md5,b3a5ec7e12cfad248f89471add43438f",
                    "versions.yml:md5,bd055fdc48b00af8df08c1f30575bfd6",
                    "versions.yml:md5,d1c67dc2d4ff10a37e0f6e53e243548e"
                ]
            },
            [
                {
                    "FASTQ_GENERATE_STATISTICS:SEQKIT_STATS": {
                        "seqkit": "2.9.0"
                    }
                },
                {
                    "FASTQ_GENERATE_STATISTICS:SEQTK_COMP": {
                        "seqtk": "1.4-r122"
                    }
                },
                {
                    "FASTQ_GENERATE_STATISTICS:FASTQC": {
                        "fastqc": "0.12.1"
                    }
                },
                {
                    "FASTQ_GENERATE_STATISTICS:SEQFU_CHECK": {
                        "seqfu": "1.22.3"
                    }
                },
                {
                    "FASTQ_GENERATE_STATISTICS:SEQFU_STATS": {
                        "seqfu": "1.22.3"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.10.0"
        },
        "timestamp": "2025-10-29T14:41:59.333124"
    },
    "sarscov2 - fastq - single_end": {
        "content": [
            "test_single_fastqc.html",
            "test_single_fastqc.zip",
            [
                [
                    {
                        "id": "test_single",
                        "single_end": true
                    },
                    "test_single.tsv:md5,4c6409169772005cfb06be9e41f2c1e2"
                ]
            ],
            [
                [
                    {
                        "id": "test_single",
                        "single_end": true
                    },
                    "test_single.tsv:md5,7573d0c83cfc9af6e1ced67a45265381"
                ]
            ],
            [
                [
                    {
                        "id": "test_single",
                        "single_end": true
                    },
                    "test_single_mqc.txt:md5,a039b8c1cc923db88d2484d3abbf00fe"
                ]
            ],
            [
                [
                    {
                        "id": "test_single",
                        "single_end": true
                    },
                    "test_single.tsv:md5,f172827a8608b646559cc39c6ca05085"
                ]
            ],
            [
                [
                    {
                        "id": "test_single",
                        "single_end": true
                    },
                    "test_single.seqtk_stats.tsv:md5,162f75dd3cfce99bd195aa5f8bd467d5"
                ]
            ],
            [
                "versions.yml:md5,5cd377fecf8c1a6bb9aa657313e882d4",
                "versions.yml:md5,7518d6130f9f32acd21e5ebaaeb02c82",
                "versions.yml:md5,b3a5ec7e12cfad248f89471add43438f",
                "versions.yml:md5,bd055fdc48b00af8df08c1f30575bfd6",
                "versions.yml:md5,d1c67dc2d4ff10a37e0f6e53e243548e"
            ]
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.10.0"
        },
        "timestamp": "2025-10-29T14:59:38.661118"
    },
    "sarscov2 - fastq - paired_end": {
        "content": [
            "test_paired_1_fastqc.html",
            "test_paired_1_fastqc.zip",
            "test_paired_2_fastqc.html",
            "test_paired_2_fastqc.zip",
            [
                [
                    {
                        "id": "test_paired",
                        "single_end": false
                    },
                    "test_paired.tsv:md5,c3e6eae028f23041bd09cac9c371f318"
                ]
            ],
            [
                [
                    {
                        "id": "test_paired",
                        "single_end": false
                    },
                    "test_paired.tsv:md5,d9277db149ae44112a704b66d72ec07e"
                ]
            ],
            [
                [
                    {
                        "id": "test_paired",
                        "single_end": false
                    },
                    "test_paired_mqc.txt:md5,082187706aad0e06b86297d4fcd60bec"
                ]
            ],
            [
                [
                    {
                        "id": "test_paired",
                        "single_end": false
                    },
                    "test_paired.tsv:md5,90cfe091d36348d467ed83ec412ba0c3"
                ]
            ],
            [
                [
                    {
                        "id": "test_paired",
                        "single_end": false
                    },
                    "test_paired.seqtk_stats.tsv:md5,162f75dd3cfce99bd195aa5f8bd467d5"
                ]
            ],
            [
                "versions.yml:md5,5cd377fecf8c1a6bb9aa657313e882d4",
                "versions.yml:md5,7518d6130f9f32acd21e5ebaaeb02c82",
                "versions.yml:md5,b3a5ec7e12cfad248f89471add43438f",
                "versions.yml:md5,bd055fdc48b00af8df08c1f30575bfd6",
                "versions.yml:md5,d1c67dc2d4ff10a37e0f6e53e243548e"
            ]
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.10.0"
        },
        "timestamp": "2025-10-29T14:59:58.678522"
    },
    "sarscov2 - fastq - paired_end - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_paired",
                            "single_end": false
                        },
                        "test_paired.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_paired",
                            "single_end": false
                        },
                        "test_paired.zip:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test_paired",
                            "single_end": false
                        },
                        "test_paired.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test_paired",
                            "single_end": false
                        },
                        "test_paired.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test_paired",
                            "single_end": false
                        },
                        "test_paired_mqc.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test_paired",
                            "single_end": false
                        },
                        "test_paired.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test_paired",
                            "single_end": false
                        },
                        "test_paired.seqtk_stats.tsv:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "7": [
                    "versions.yml:md5,5cd377fecf8c1a6bb9aa657313e882d4",
                    "versions.yml:md5,7518d6130f9f32acd21e5ebaaeb02c82",
                    "versions.yml:md5,b3a5ec7e12cfad248f89471add43438f",
                    "versions.yml:md5,bd055fdc48b00af8df08c1f30575bfd6",
                    "versions.yml:md5,d1c67dc2d4ff10a37e0f6e53e243548e"
                ],
                "fastqc_html": [
                    [
                        {
                            "id": "test_paired",
                            "single_end": false
                        },
                        "test_paired.html:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "fastqc_zip": [
                    [
                        {
                            "id": "test_paired",
                            "single_end": false
                        },
                        "test_paired.zip:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "seqfu_check": [
                    [
                        {
                            "id": "test_paired",
                            "single_end": false
                        },
                        "test_paired.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "seqfu_multiqc": [
                    [
                        {
                            "id": "test_paired",
                            "single_end": false
                        },
                        "test_paired_mqc.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "seqfu_stats": [
                    [
                        {
                            "id": "test_paired",
                            "single_end": false
                        },
                        "test_paired.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "seqkit_stats": [
                    [
                        {
                            "id": "test_paired",
                            "single_end": false
                        },
                        "test_paired.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "seqtk_stats": [
                    [
                        {
                            "id": "test_paired",
                            "single_end": false
                        },
                        "test_paired.seqtk_stats.tsv:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,5cd377fecf8c1a6bb9aa657313e882d4",
                    "versions.yml:md5,7518d6130f9f32acd21e5ebaaeb02c82",
                    "versions.yml:md5,b3a5ec7e12cfad248f89471add43438f",
                    "versions.yml:md5,bd055fdc48b00af8df08c1f30575bfd6",
                    "versions.yml:md5,d1c67dc2d4ff10a37e0f6e53e243548e"
                ]
            },
            [
                {
                    "FASTQ_GENERATE_STATISTICS:SEQKIT_STATS": {
                        "seqkit": "2.9.0"
                    }
                },
                {
                    "FASTQ_GENERATE_STATISTICS:SEQTK_COMP": {
                        "seqtk": "1.4-r122"
                    }
                },
                {
                    "FASTQ_GENERATE_STATISTICS:FASTQC": {
                        "fastqc": "0.12.1"
                    }
                },
                {
                    "FASTQ_GENERATE_STATISTICS:SEQFU_CHECK": {
                        "seqfu": "1.22.3"
                    }
                },
                {
                    "FASTQ_GENERATE_STATISTICS:SEQFU_STATS": {
                        "seqfu": "1.22.3"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.10.0"
        },
        "timestamp": "2025-10-29T14:42:13.823178"
    }
}